api/scholarqa/scholar_qa.py

```diff
@@ -609,3 +613,422 @@ def run_qa_pipeline(self, req: ToolRequest, inline_tags=False) -> TaskResult:
         event_trace.trace_summary_event(json_summary, all_sections, tcosts)
         event_trace.persist_trace(self.logs_config)
         return TaskResult(report_title=self.report_title, sections=generated_sections, cost=event_trace.total_cost, tokens=event_trace.tokens)
+
+    @traceable(run_type="tool", name="ai2_scholar_qa_edit_trace")
+    def run_edit_pipeline(self, req: ToolRequest, inline_tags=False) -> TaskResult:
+        """
+        Edit an existing report based on user instructions.
+
+        This function MIRRORS run_qa_pipeline with edit-specific extensions:
+        0) Retrieve current report from thread_id
+        1) Conditional search + rerank (includes mentioned_papers)
+        2) Quote extraction from new papers (WITH EDIT CONTEXT)
+        3) Edit planning/clustering (WITH CURRENT REPORT)
+        4) Section generation/editing (WITH CURRENT SECTIONS + ACTIONS)
+
+        :param req: Tool request with edit_existing=True, thread_id, edit_instruction, mentioned_papers
+        :param inline_tags: Whether to include inline <paper> tags in the output
+        :return: Updated TaskResult
+        """
+        self.tool_request = req
+        self.update_task_state("Processing edit request", task_estimated_time="~3 minutes", step_estimated_time=5)
+
+        # Validate edit request
+        if not req.edit_instruction:
+            raise ValueError("edit_instruction is required for edit workflow")
+        if not req.thread_id:
+            raise ValueError("thread_id is required to fetch the current report")
+
+        task_id = self.task_id if self.task_id else req.task_id
+        user_id, msg_id = self.get_user_msg_id()
+        msg_id = task_id if not msg_id else msg_id
+        edit_instruction = req.edit_instruction
+
+        logger.info(
+            f"Received edit request for thread {req.thread_id}: {edit_instruction} "
+            f"from user_id: {user_id}"
+        )
+
+        # ====================================================================
+        # STEP 0: Retrieve current report (EDIT-SPECIFIC)
+        # ====================================================================
+        self.update_task_state("Retrieving current report", step_estimated_time=2)
+        current_report = self._retrieve_report_from_thread(req.thread_id)
+        if not current_report:
+            raise ValueError(f"No report found for thread_id: {req.thread_id}")
+
+        logger.info(
+            f"Retrieved report with {len(current_report.sections)} sections: "
+            f"{current_report.report_title}"
+        )
+
+        # Keep original report title
+        self.report_title = current_report.report_title
+
+        # Initialize event trace
+        event_trace = EventTrace(
+            task_id,
+            self.paper_finder.retriever.n_retrieval if hasattr(self.paper_finder.retriever, "n_retrieval") else 0,
+            self.paper_finder.n_rerank,
+            req,
+            user_id=user_id
+        )
```

Comment:
probably need to think of a better way to log traces, or at the very least map the new traces to older ones